Port xmbcclient.py to Python 3, retaining Python 2 compatibility

chrisjbillington opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
This change ports the xbmcclient Python library to Python 3, retaining Python 2 compatibility by using the language features that allow the same codebase to function on both versions of Python.
This mostly involved `__future__` imports to turn on several Python 3 features in Python 2, aliasing `str = unicode` in Python 2, and then ensuring strings that needed to be bytestrings were either defined as such using the `b'string'` syntax, or encoded to bytestrings before concatenation with other bytestrings.

Also removed a bare except and changed it to raise the exception if packet sending fails: eating the exception and returning False is considered poor form in Python as it is not helpful to calling code in knowing what went wrong - calling code can catch the exception if it expects something in particular to go wrong. Since this is a behavioural change I will revert it if desired, but I would guess there is approximately zero code relying on this behaviour.

## Motivation and Context

This allows code written in Python 3 to make calls to the xmbcclient library for controlling kodi. This is useful since an increasing amount of code is written in Python 3. And there is little downside since Python 2 compatibility can be retained. 

## How Has This Been Tested?

I have tested that kodi-send works unchanged to send a 'mute' action on both Python 2.7 and Python 3.6, and I have tested that my script (Python 3.6) for sending actions to kodi also works (it does the same sort of thing as kodi-send). I have tested that all the examples in `tools/EventClients/examples/python` still work in Python 2.7 and on Python 3.6 (they all work on Python 3.6 unchanged). I have not attempted to port zeroconf.py to Python 3 though I have glanced over it and there is nothing obvious that needs to be changed.

Unsure if documentation changes are required for this at this point, please advise.

Another change necessary for users to actually use the library from Python 3 will be to get it into the Python 3 package dirs, which I am unsure how to do.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

chrisjbillington commented:
No worries!

I made the port so that I could use xbmcclient.py with Python3 for a [tool I've made](https://bitbucket.org/cbillington/kodi_de_diplomat/) to address some volume and media key input issues that linux users may experience running kodi as well as an ordinary desktop environment on the same box. For the moment I'm just bundling this Python 3 port of xbmcclient.py with my tool, but will remove it after the next kodi release if this PR is accepted.

tadly commented:
Thanks for the contribution.
Reading through the changes, this looks good to me :)

On a side note I think we could do better in terms of formatting in this file to be more PEP8 compliant.
This does **not** affect this PR, just something I noticed while reading.

chrisjbillington commented:
> Thanks for the contribution.

You're welcome, thanks for reviewing :)

> I think we could do better in terms of formatting in this file to be more PEP8 compliant

I noticed that but didn't touch any except the most egregious (two spaces for indentation in places)

tadly commented on tools/EventClients/lib/python/xbmcclient.py:
```diff
@@ -249,10 +258,7 @@ def send(self, sock, addr, uid=UNIQUE_IDENTIFICATION):
         """
         self.uid = uid
         for a in range ( 0, self.num_packets() ):
-            try:
```
I'd actually like to discuss this change after thinking about it some more.
I do agree with your reasoning but:
 1. that `return True` is now obsolet.
 2. more importantly, there's `kodi-send` which does rely on this and would spew python exception to the user in case of failed connections. Do we want this? Do we care? I'm not sure.

**Edit:**
P.s. No clue where to find the source for the `kodi-send` script.

tadly replied:
Thanks @hudokkow 
I assume that the comment of mine where I said "Found it" is only visible to me for whatever reason.
I'm so confused by github right now but anyways... For I shall repeat myself :)

[This](https://github.com/xbmc/xbmc/blob/c49d783988a6c1c3adaabfe1425f42e85d3a5ef1/tools/EventClients/Clients/Kodi%20Send/kodi-send.py#L76-L77) is what I was referring to specifically. (Let's not talk about how this file does `from ... import *`)

chrisjbillington replied:
Ah so, when I said approximately zero code is relying on this behaviour, I meant zero code is checking the return value and doing anything more useful with it than raising an exception. Sending a UDP packet can't really fail unless the process has too many file descriptors, or the user doesn't have permissions, or weird things like that. Any exceptions being supressed by this are likely bugs in kodi-send or xbmcclient.py and should be raised as dirty tracebacks just like if any other part of the code failed.

Since kodi-send was not checking the return value, the only change in its behaviour is that if there is a bug in xbmcclient.py or something going wrong with the operating system like ephemeral port exhaustion, it will print a traceback and give a nonzero exit code. This is probably desirable, and is what kodi-send already does in the case of unlikely failures in other parts of the code since exceptions are not suppressed anywhere else.

I actually removed the try: except: in the first place because it was masking bugs in my porting to Python 3.

tadly replied:
Guess who totally ignored the fact we're talking upd here  _points at self_
Sounds all reasonable and I'm all good :)

tadly reviewed (commented):
[Found it](https://github.com/xbmc/xbmc/blob/c49d783988a6c1c3adaabfe1425f42e85d3a5ef1/tools/EventClients/Clients/Kodi%20Send/kodi-send.py#L76-L77) :)

That was supposed to go to the review, not into a separate comment but I can't delete it so  ¯\_(ツ)_/¯

hudokkow commented:
Ask, and it shall be given you... ;)

https://github.com/xbmc/xbmc/blob/master/tools/EventClients/Clients/Kodi%20Send/kodi-send.py

chrisjbillington commented:
Removed redundant return statements - since other send() methods were returning the return value of Packet.send(), these were also redundant.

I typod 'remobed' instead of 'removed' in the commit message, can fix if this is considered important.

hudokkow commented on tools/EventClients/lib/python/xbmcclient.py:
```diff
@@ -37,11 +37,16 @@
+    str = unicode
 from struct import pack
-from socket import *
+from socket import socket, AF_INET,SOCK_DGRAM, SOL_SOCKET, SO_BROADCAST
```
Missing space before `SOCK_DGRAM`

enen92 commented on tools/EventClients/lib/python/xbmcclient.py:
```diff
@@ -37,11 +37,16 @@
+from __future__ import unicode_literals, print_function, absolute_import, division
+
 __author__  = "d4rk@xbmc.org"
 __version__ = "0.0.3"
```
Maybe the version should also be bumped?

enen92 commented on tools/EventClients/lib/python/xbmcclient.py:
```diff
@@ -279,7 +282,7 @@ def __init__(self, devicename=None, icon_type=ICON_NONE, icon_file=None):
         self.append_payload( format_uint32 (0) ) # reserved2
         if icon_type != ICON_NONE and icon_file:
-            self.append_payload( file(icon_file).read() )
+            self.append_payload( open(icon_file, 'rb').read() )
```
Probably better to use python's `with` statement here to ensure the file is always closed

enen92 commented on tools/EventClients/lib/python/xbmcclient.py:
```diff
@@ -305,7 +308,7 @@ def __init__(self, title, message, icon_type=ICON_NONE, icon_file=None):
         self.append_payload( format_uint32 (0) ) # reserved
         if icon_type != ICON_NONE and icon_file:
-            self.append_payload( file(icon_file).read() )
+            self.append_payload( open(icon_file, 'rb').read() )
```
same here

enen92 commented on tools/EventClients/lib/python/xbmcclient.py:
```diff
@@ -590,13 +593,13 @@ def send_button_state(self, map="", button="", amount=0, down=0, axis=0):
-            down = 0
-          else:
-            down = 1
+            if amount == 0:
```
`down = int(ammount != 0)` would suffice but it is fine as it is

enen92 reviewed (approved):
LGTM (see comments)

chrisjbillington commented:
Thanks for the reviews, I've made the suggested changes and tested that all the examples still work.

Razzeee commented:
jenkins build this please

chrisjbillington committed:
Port xbmcclient.py to Python 3, retaining Python 2 compatibility

Bump minor version to 0.1 and remove blanket exception catching
in favour of raising unexpected exceptions.

chrisjbillington commented:
I've squashed the lot into one commit, is this better?

Razzeee commented:
yes, it's better

jenkins build this please

Razzeee commented:
Thank you @chrisjbillington

chrisjbillington commented:
Thanks!
